### PR TITLE
Fix AWS app access for Java SDK v1

### PR DESCRIPTION
**Pull request by greedy52**
Fixes #19366

Changes:
- Replace AWS signature verification with AWS AccessKeyID verification at tsh local proxy, when "User-Agent" header is detected as part of the "SignedHeaders")

Security:
Verifying requests received by the local proxy is optional. Instead of just skipping the whole signature verification, we now still check if the requests have a signature header, and verifies the AccessKeyID is the expected one. A new UUID is generated for each tsh command to avoid a constant AccessKeyID.

Will add jdbc driver support for Athena and Timestream separately.

### CODE REVIEW

**Review by GavinFrazar — APPROVED** [review]

File: lib/srv/alpnproxy/aws_local_proxy.go [hunk 1]
```diff
@@ -47,8 +47,14 @@ func (m *AWSAccessMiddleware) CheckAndSetDefaults() error {
 }
 
 func (m *AWSAccessMiddleware) HandleRequest(rw http.ResponseWriter, req *http.Request) bool {
-	if err := aws.VerifyAWSSignature(req, m.AWSCredentials); err != nil {
-		m.Log.WithError(err).Error("AWS signature verification failed.")
+	// It's found that some AWS clients may have slightly different signing
+	// logic than the Go SDK. For example, the Java SDK v1 includes the
+	// "Use-Agent" header when signing but the Go SDK explicitly excludes this
+	// header. To workaround, only AccessKeyID from the signature is verified
```
Comment (codingllama) [inline lib/srv/alpnproxy/aws_local_proxy.go:53]: Could we try to verify both with and without the User-Agent? Would it provide any hints to whether we are seeing a request from Go or Java?

File: lib/srv/alpnproxy/aws_local_proxy.go [hunk 2]
```diff
@@ -47,8 +47,14 @@ func (m *AWSAccessMiddleware) CheckAndSetDefaults() error {
 }
 
 func (m *AWSAccessMiddleware) HandleRequest(rw http.ResponseWriter, req *http.Request) bool {
-	if err := aws.VerifyAWSSignature(req, m.AWSCredentials); err != nil {
-		m.Log.WithError(err).Error("AWS signature verification failed.")
+	// It's found that some AWS clients may have slightly different signing
+	// logic than the Go SDK. For example, the Java SDK v1 includes the
+	// "Use-Agent" header when signing but the Go SDK explicitly excludes this
```
Comment (codingllama) [inline lib/srv/alpnproxy/aws_local_proxy.go:52]:
```suggestion
	// "User-Agent" header when signing but the Go SDK explicitly excludes this
```

File: lib/utils/aws/aws.go [hunk 3]
```diff
@@ -171,23 +172,38 @@ func tryDrainBody(b io.ReadCloser) (payload []byte, err error) {
...
+// VerifyAWSAccessKeyID verifies the AccessKeyID from the request signature.
+func VerifyAWSAccessKeyID(req *http.Request, credentials *credentials.Credentials) error {
+	_, err := parseSigV4AndVerifyAccessKeyID(req, credentials)
+	return trace.Wrap(err)
+}
+
+// VerifyAWSSignature verifies the request signature ensuring that the request originates from tsh aws command execution
+// AWS CLI signs the request with random generated credentials that are passed to LocalProxy by
+// the AWSCredentials LocalProxyConfig configuration.
+func VerifyAWSSignature(req *http.Request, credentials *credentials.Credentials) error {
```
Comment (codingllama) [inline lib/utils/aws/aws.go:203]: Is this function ever called by non-testing code after the change?

File: lib/srv/alpnproxy/aws_local_proxy.go [hunk 4]
```diff
@@ -47,8 +47,14 @@ func (m *AWSAccessMiddleware) CheckAndSetDefaults() error {
...
+	// header. To workaround, only AccessKeyID from the signature is verified
+	// here. A new UUID should be generated as the AccessKeyID per tsh command
+	// for better security.
+	if err := aws.VerifyAWSAccessKeyID(req, m.AWSCredentials); err != nil {
```
Comment (codingllama) [inline lib/srv/alpnproxy/aws_local_proxy.go:56]: If we drop signature validation, does that mean that the AccessKeyID is the only data one needs to get past this check? Isn't this weakening the properties of the system?

File: tool/tsh/aws.go [hunk 5]
```diff
@@ -364,6 +348,71 @@ func (a *awsApp) startLocalForwardProxy(port string) error {
...
 }
 
+// SetupKeyStore installs local CA to the specified Java KeyStore.
+func (a *awsApp) SetupKeyStore(keystore string) error {
+	aliasName := fmt.Sprintf("teleport-%v-%v", a.profile.Cluster, a.appName)
+
+	// Print info like alias name in case user wants to manually update the
+	// KeyStore. Use new lines between keytool commands as it may prompt user
+	// for password or print some outputs.
+	fmt.Fprintln(a.cf.Stdout(), "KeyStore path: ", keystore)
```
Comment (codingllama) [inline tool/tsh/aws.go:358]: My 2c: all of those should be debug logs.

File: tool/tsh/aws.go [hunk 6]
```diff
@@ -364,6 +348,71 @@ func (a *awsApp) startLocalForwardProxy(port string) error {
 	return nil
 }
 
+// SetupKeyStore installs local CA to the specified Java KeyStore.
```
Comment (codingllama) [inline tool/tsh/aws.go:351]: What is the format of a "Java Keystore"? Is it PKCS#12? Proprietary?

File: tool/tsh/aws.go [hunk 7]
```diff
@@ -364,6 +348,71 @@ func (a *awsApp) startLocalForwardProxy(port string) error {
...
+		return trace.Wrap(err)
+	}
+
+	fmt.Fprintln(a.cf.Stdout())
+	return nil
+}
+
+func (a *awsApp) deleteLocalCAFromKeyStore(keystore, aliasName string) error {
+	cmd := exec.Command(
+		"keytool", "-noprompt", "-delete",
```
Comment (codingllama) [inline tool/tsh/aws.go:382]: Could we do this without the external dependency? This looks easy to break in a variety of systems.

File: tool/tsh/aws.go [hunk 8]
```diff
@@ -364,6 +348,71 @@ func (a *awsApp) startLocalForwardProxy(port string) error {
...
+	err := a.cf.RunCommand(cmd)
+	if err != nil && !strings.Contains(capture.String(), fmt.Sprintf("Alias <%s> does not exist", aliasName)) {
+		return trace.Wrap(err)
+	}
+	return nil
+}
+
+func (a *awsApp) installLocalCAToKeyStore(keystore, aliasName string) error {
+	cmd := exec.Command(
+		"keytool", "-noprompt", "-import",
```
Comment (codingllama) [inline tool/tsh/aws.go:404]: Same.

File: tool/tsh/aws.go [hunk 9]
```diff
@@ -364,6 +348,71 @@ func (a *awsApp) startLocalForwardProxy(port string) error {
...
+		"-alias", aliasName,
+		"-keystore", keystore,
+		"-file", a.profile.AppLocalCAPath(a.appName),
+	)
+	cmd.Stdout = a.cf.Stdout()
+	cmd.Stderr = a.cf.Stderr()
+	cmd.Stdin = a.cf.Stdin()
+
+	return trace.Wrap(a.cf.RunCommand(cmd))
+}
```
Comment (codingllama) [inline tool/tsh/aws.go:414]: Let's move the keystore-related logic outside of ./tool/tsh and to its own package, where we can test it independently as well.

File: tool/tsh/tsh.go [hunk 10]
```diff
@@ -341,6 +341,8 @@ type CLIConf struct {
 	// AWSEndpointURLMode is an AWS proxy mode that serves an AWS endpoint URL
 	// proxy instead of an HTTPS proxy.
 	AWSEndpointURLMode bool
+	// KeyStore is the path to a Java KeyStore.
```
Comment (codingllama) [inline tool/tsh/tsh.go:344]: Please call out the format of the keystore here too.

File: tool/tsh/tsh.go [hunk 11]
```diff
@@ -704,6 +707,7 @@ func Run(ctx context.Context, args []string, opts ...cliOption) error {
 	proxyAWS.Flag("port", "Specifies the source port used by the proxy listener.").Short('p').StringVar(&cf.LocalProxyPort)
 	proxyAWS.Flag("endpoint-url", "Run local proxy to serve as an AWS endpoint URL. If not specified, local proxy serves as an HTTPS proxy.").Short('e').BoolVar(&cf.AWSEndpointURLMode)
 	proxyAWS.Flag("format", awsProxyFormatFlagDescription()).Short('f').Default(envVarDefaultFormat()).EnumVar(&cf.Format, awsProxyFormats...)
+	proxyAWS.Flag("keystore", "Path to a Java KeyStore. If specified, the local CA will be imported to this KeyStore.").StringVar(&cf.KeyStore)
```
Comment (codingllama) [inline tool/tsh/tsh.go:710]: Please call out the format of the keystore here too.

**Review by codingllama — COMMENTED** [review]
A couple of high-level remarks:

1. Dropping signature validation makes me a bit uneasy, although it may just be that I don't understand the security properties it gives in the first place.

2. Binding tsh to `keytool` seems liable to break. I think we should either do it in pure Go, or tell Java keystore users how to transform proprietary formats to ones we can handle (handling an open format seems more valuable in the long-run anyway).

File: lib/srv/alpnproxy/aws_local_proxy.go (hunk 1 above)
Comment (greedy52) [reply lib/srv/alpnproxy/aws_local_proxy.go:53]: > Could we try to verify both with and without the User-Agent? 
Some of the signature logic including [`ignoredHeaders` ](https://github.com/aws/aws-sdk-go/issues/4662) is private and hard-coded in the Go SDK. So we cannot try to sign "with" it.

> Would it provide any hints to whether we are seeing a request from Go or Java?
It's kinda possible. We know what headers are included in the signature. So technically we can verify full when 'User-Agent' is not present there, and only do the AccessKeyID way only when 'User-Agent' is present. Do you think that would be better? My thought process was if we don't verify the whole thing for some cases, then why bother doing it for others. But I am open to preserving the full check if we think it's better for security.

This also ties to your question about dropping the signature validation. As mentioned in the description, the local validation is optional at best. We don't do validations for other local proxies (e.g `tsh proxy db --tunnel`).

File: tool/tsh/aws.go (hunk 6 above)
Comment (greedy52) [reply tool/tsh/aws.go:351]: It should be PKCS12 by default starting Java 9. Though, I couldn't get `openssl pkcs12` on my JDK's default KeyStore.

File: tool/tsh/aws.go (hunk 7 above)
Comment (greedy52) [reply tool/tsh/aws.go:382]: `keytool` is specific to Java and is included in JDK. So if you are using AWS Java SDKs, the binary should be there. The new `--key-store` is an optional flag that tries to help users setup. But today users run `tsh proxy aws` for Java SDK v2 without this flag anyway. 

I see a few golang implementations of `keytool` like [minijks](https://github.com/lwithers/minijks) but it feels not very good.

I don't think PKCS12 will always work out either. I think `keytool` is the best bet.

Comment (greedy52) [reply tool/tsh/aws.go:382]: I guess one other alternative would be just printing out instructions for the users to consider and execute separately.

File: tool/tsh/tsh.go (hunk 10 above)
Comment (greedy52) [reply tool/tsh/tsh.go:344]: As mentioned above, I would consider `KeyStore` is the format.

Comment (greedy52) [reply tool/tsh/tsh.go:344]: I think maybe the confusion here is I only want to support the [Java KeyStore](https://docs.oracle.com/javase/7/docs/api/java/security/KeyStore.html), not any generic keystore?

File: tool/tsh/aws.go (hunk 6 above)
Comment (codingllama) [reply tool/tsh/aws.go:351]: Let's try to name it and call it out the best we can, this kind of info is helpful.

Comment (codingllama) [reply tool/tsh/aws.go:351]: A cursory search led me to this: https://docs.oracle.com/javase/8/docs/technotes/guides/security/StandardNames.html#KeyStore (it's pretty old, though).

If it's jks or jceks, let's spell it out.

File: tool/tsh/aws.go (hunk 7 above)
Comment (codingllama) [reply tool/tsh/aws.go:382]: I think we could consider the following:

1. Let users specify the location of keytool via an env var (default to $PATH search)
2. If we can't find keytool, we print instructions explaining the env var and/or the command they may run.

WDYT?

File: tool/tsh/tsh.go (hunk 10 above)
Comment (codingllama) [reply tool/tsh/tsh.go:344]: Same as the other comment, let's call out whether this is jks, jceks, both, etc.

File: lib/srv/alpnproxy/aws_local_proxy.go (hunk 1 above)
Comment (codingllama) [reply lib/srv/alpnproxy/aws_local_proxy.go:53]: > Some of the signature logic including https://github.com/aws/aws-sdk-go/issues/4662is private and hard-coded in the Go SDK. So we cannot try to sign "with" it.

I thought you might say that.

> So technically we can verify full when 'User-Agent' is not present there, and only do the AccessKeyID way only when 'User-Agent' is present

If signature validation serves a purpose, I think I'd rather do it only for Go than not at all. If it doesn't serve a purpose, then I'd be OK cutting it out. I don't understand the overall picture to make a call, though.

File: tool/tsh/aws.go (hunk 7 above)
Comment (greedy52) [reply tool/tsh/aws.go:382]: Sounds good! I think we use `exec.LookPath` in some other places. We can do that first on `keytool` and if that fails we can just print the instructions.

File: lib/srv/alpnproxy/aws_local_proxy.go (hunk 1 above)
Comment (greedy52) [reply lib/srv/alpnproxy/aws_local_proxy.go:53]: Ok. Then let me update this part to only do AccessKeyID validation when 'User-Agent' is getting signed. So everything else is same as before

File: tool/tsh/aws.go (hunk 6 above)
Comment (greedy52) [reply tool/tsh/aws.go:351]: Personally, I don't think we need to know the format. Different versions of Java may use different formats but the `keytool` included in the install should always match/work with the KeyStore (at least the default) the Java developers will be using for their programs.

The `keytool` does support other formats through the `-storetype` flag though.

What do you think we call it `JavaKeyStore` here and I will rename the path flag to`--java-keystore`, but I will also add an extra flag to this tsh command like `--java-keystore-type` and default that to `PKCS12`? That way it's clear what format is going to be used for the keystore and users also get an option to change it if needed.

Comment (codingllama) [reply tool/tsh/aws.go:351]: The problem with saying "JavaKeyStore" is that it's not a format, but a bag of varying implementations. It sounds like by "JavaKeyStore" we actually mean "keystores that keytool is able to manage", which is still open-ended but a bit more precise.

My suggestion is to sum up the changes at a high level and take it to product/UX for feedback, so we are not making UX decisions in isolation. There's some nuance here on how to find and handle `keytool`, `keytool --storetype`, whether we should even do this in tsh or just print instructions, etc.

Comment (greedy52) [reply tool/tsh/aws.go:351]: Sounds good. I will drop that part then from this PR and review it with the product team. I will update the verify part so we can get that fix first.

**Review by codingllama — APPROVED** [review]

**Review by smallinsky — APPROVED** [review]

**Pull request merged**